File: app/src/main/java/com/lmgy/exportapk/utils/SearchTask.java
package com.lmgy.exportapk.utils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.lmgy.exportapk.Global;
import com.lmgy.exportapk.bean.AppItemBean;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * @author lmgy
 * @date 2019/10/13
 */
public class SearchTask implements Runnable {

    private boolean isInterrupted = false;
    private final String searchInfo;
    private final List<AppItemBean> totalList;
    private final ArrayList<AppItemBean> searchResult = new ArrayList<>();
    private SearchTaskCompletedCallback callback;

    public SearchTask(@NonNull List<AppItemBean> totalList, @NonNull String info, @Nullable SearchTaskCompletedCallback callback) {
        this.searchInfo = info;
        this.totalList = totalList;
        this.callback = callback;
    }


    @Override
    public void run() {
        for (AppItemBean item : totalList) {
            if (isInterrupted) {
                searchResult.clear();
                return;
            }
            try {
                boolean b = (getFormatString(item.getAppName()).contains(searchInfo)
                        || getFormatString(item.getPackageName()).contains(searchInfo)
                        || getFormatString(item.getVersionName()).contains(searchInfo)
                        || PinyinUtils.getFirstSpell(item.getAppName()).contains(searchInfo)
                        || PinyinUtils.getFullSpell(item.getAppName()).contains(searchInfo)
                        || PinyinUtils.getPinYin(item.getAppName()).contains(searchInfo)) && !"".equals(searchInfo.trim());
                if (b) {
                    searchResult.add(item);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        Global.HANDLER.post(() -> {
            if (callback != null && !isInterrupted) {
                callback.onSearchTaskCompleted(searchResult);
            }
        });
    }

    public void setInterrupted() {
        isInterrupted = true;
    }

    @NonNull
    private static String getFormatString(@NonNull String s) {
        return s.trim().toLowerCase(Locale.getDefault());
    }

    public interface SearchTaskCompletedCallback {
        void onSearchTaskCompleted(@NonNull List<AppItemBean> result);
    }

}

File: app/src/main/java/com/lmgy/exportapk/utils/SearchTask.kt
package com.lmgy.exportapk.utils

import com.lmgy.exportapk.base.BaseActivity
import com.lmgy.exportapk.bean.AppItemBean
import java.util.*

/**
 * @author lmgy
 * @date 2019/10/13
 */
class SearchTask(searchInfo: String) : Runnable {

    private val searchInfo: String
    private var ifCanSearch: Boolean = false
    private val listSearch: MutableList<AppItemBean>
    private val listSum: List<AppItemBean>

    init {
        this.searchInfo = searchInfo.trim { it <= ' ' }.toLowerCase(Locale.ENGLISH)
        this.ifCanSearch = true
        this.listSum = BaseActivity.listSum
        this.listSearch = ArrayList()
    }

    override fun run() {
        if (this.listSum != null) {
            listSearch.clear()
            if (searchInfo.isNotEmpty()) {
                for (i in this.listSum.indices) {

                    if (this.ifCanSearch) {
                        try {
                            if (this.listSum[i].appName!!.toLowerCase(Locale.ENGLISH).contains(searchInfo)
                                    || this.listSum[i].packageName!!.toLowerCase(Locale.ENGLISH).contains(searchInfo)
                                    || this.listSum[i].version.toLowerCase(Locale.ENGLISH).contains(searchInfo)
                                    || PinyinUtils.getFullSpell(this.listSum[i].appName!!).toLowerCase(Locale.ENGLISH).contains(searchInfo)
                                    || PinyinUtils.getFirstSpell(this.listSum[i].appName!!).toLowerCase(Locale.ENGLISH).contains(searchInfo)
                                    || PinyinUtils.getPinYin(this.listSum[i].appName!!).toLowerCase(Locale.ENGLISH).contains(searchInfo)) {
                                listSearch.add(this.listSum[i])
                            }
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }

                    } else {
                        listSearch.clear()
                        break
                    }
                }
            } else {
                listSearch.clear()
            }
        } else {
            this.ifCanSearch = false
        }
        if (this.ifCanSearch) {
            BaseActivity.listSearch = this.listSearch
            BaseActivity.sendEmptyMessage(Main.MESSAGE_SEARCH_COMPLETE)
        }
    }

    fun setInterrupted() {
        this.ifCanSearch = false
    }

}
